cub/cub/util_type.cuh

```diff
@@ -942,18 +899,15 @@ struct DoubleBuffer
  * \brief Defines a structure \p detector_name that is templated on type \p T.  The \p detector_name struct exposes a
  * constant member \p VALUE indicating whether or not parameter \p T exposes a nested type \p nested_type_name
  */
-#  define CUB_DEFINE_DETECT_NESTED_TYPE(detector_name, nested_type_name) \
-    template <typename T>                                                \
-    struct detector_name                                                 \
-    {                                                                    \
-      template <typename C>                                              \
-      static char& test(typename C::nested_type_name*);                  \
-      template <typename>                                                \
-      static int& test(...);                                             \
-      enum                                                               \
-      {                                                                  \
-        VALUE = sizeof(test<T>(0)) < sizeof(int)                         \
-      };                                                                 \
+#  define CUB_DEFINE_DETECT_NESTED_TYPE(detector_name, nested_type_name)                                  \
+    template <typename T, typename = void>                                                                \
+    struct detector_name : ::cuda::std::false_type                                                        \
+    {                                                                                                     \
+      CUB_DEPRECATED_BECAUSE("Use ::value instead") static constexpr bool VALUE = false;                  \
+    };                                                                                                    \
+    struct detector_name<T, ::cuda::std::__void_t<typename C::nested_type_name>> : ::cuda::std::true_type \
+    {                                                                                                     \
+      CUB_DEPRECATED_BECAUSE("Use ::value instead") static constexpr bool VALUE = true;                   \
```

Comment:
important: if you want to change this code, consider adding a test first

Reply:
I wonder whether I shouldn't just deprecate the whole macro. It's not used in our code base.

I added a test and of course found a bug. Thx! :D